### PR TITLE
Cleaned up write_command

### PR DESCRIPTION
**Pull request by couger01**
* Stripped newline and returned list for methods using write_command
* Changed decoding to use ISO-8959-1 instead of utf-8
* Changed methods to account for new return type
* Changed multi_line to num_line so that the expected number of lines is easier
  to understand

### CODE REVIEW

File: python/lsst/ts/ATHexapod/controller.py [hunk 1]
```diff
@@ -121,22 +122,24 @@ def is_connected(self):
...
-                        elif c == b'\n':
-                            keep_going = False
                 except asyncio.TimeoutError:
                     self.log.warning("Timed out waiting for response from controller. Result "
                                      "may be incomplete.")
 
         self.log.debug(f"Response: {response!r}")
+        response = response.decode("ISO-8859-1")
+        response = response.replace("\n", "")
+        response = response.split(" ")
```
Comment (r-owen) [inline python/lsst/ts/ATHexapod/controller.py]: Based on some of your test code below I am guessing that the spaces your code deals with are just leading spaces that are meaningless. For example `f"X=1\n Y=1\n Z=1\n U=1\n V=1\n W=1\n"`. If that is the case then please consider using `strip` to get rid of them. I think the following code does everything you need:
```
replies = []
for i in range(num_line):
    raw_line = await self.reader.readuntil(b'\n')
    line = raw_line.decode("ISO-8859-1").strip()  # Remove trailing \n and leading whitespace
    self.log.debug(f"Read {i+1} of {num_line} lines: {line}")
    replies.append(line)
```
This will need work if num_line can be 0, but I didn't find any such calls.

Comment (r-owen) [reply python/lsst/ts/ATHexapod/controller.py]: I strongly suggest raising an exception if the read times out. That makes the calling code simpler because it can assume it gets the right number of replies. I admit it may require trapping that exception somewhere up calling chain, unless you can just ignore it and log the error in `write_command`.

File: python/lsst/ts/ATHexapod/controller.py [hunk 2]
```diff
@@ -208,9 +211,9 @@ def is_connected(self):
             The value of each axis changed or not.
 
         """
-        ret = await self.write_command("\6", multi_line=False)
+        ret = await self.write_command("\6", num_line=1)
 
-        code = int(ret.decode())
+        code = int(ret)
```
Comment (r-owen) [inline python/lsst/ts/ATHexapod/controller.py]: ret is a list, so I think you need `int(ret[0])`.

File: python/lsst/ts/ATHexapod/controller.py [hunk 3]
```diff
@@ -228,7 +231,7 @@ def is_connected(self):
         comp : `str`
             A character indicating the controller is ready or not ready.
         """
-        ret = await self.write_command("\7", multi_line=False)
+        ret = await self.write_command("\7", num_line=1)
```
Comment (r-owen) [inline python/lsst/ts/ATHexapod/controller.py:247]: The code below assumes `ret` is a string.

File: python/lsst/ts/ATHexapod/mock_server.py [hunk 4]
```diff
@@ -278,7 +278,7 @@ def __init__(self, log=None):
 
     async def check_offset(self, x, y, z, u, v, w):
         """Check that the hexapod can move."""
-        return (f"X=1 Y=1 Z=1 U=1 V=1 W=1\n")
+        return (f"X=1\n Y=1\n Z=1\n U=1\n V=1\n W=1\n")
```
Comment (r-owen) [inline python/lsst/ts/ATHexapod/mock_server.py:304]: Does the real hexapod actually output this, or the old  string?

**Review by r-owen — CHANGES_REQUESTED** [review]
I found a few small bugs (code that expects the return value from `write_command` to be a string instead of a list).

Also I strongly recommend raising an exception in `write_command` if it cannot read the required # of lines.

Finally, please consider simplifying the code in `write_command` (example provided).

File: python/lsst/ts/ATHexapod/controller.py [hunk 5]
```diff
@@ -97,13 +98,13 @@ def is_connected(self):
             Command to send to hexapod. A 'newline' character will be appended.
         has_response : `bool`
             Does the command have a response?
-        multi_line : `bool`
-            Is the response received in multiple lines?
+        num_line : `int`
+            The number of expected lines.
```
Comment (r-owen) [inline python/lsst/ts/ATHexapod/controller.py]: Is this also the number of expected replies? I hope it is. Please say so in this doc string.

File: python/lsst/ts/ATHexapod/mock_server.py (hunk 4 above)
Comment (couger01) [reply python/lsst/ts/ATHexapod/mock_server.py:304]: This function is meant to "replicate" the virtual move check that the hexapod does in order to make sure that the move works in real life. However, I don't actually validate the movement in the mock server, so this just assumes everything's hunky dory.

File: python/lsst/ts/ATHexapod/controller.py [hunk 6]
```diff
@@ -97,13 +121,13 @@ def is_connected(self):
             Command to send to hexapod. A 'newline' character will be appended.
         has_response : `bool`
             Does the command have a response?
-        multi_line : `bool`
-            Is the response received in multiple lines?
+        num_line : `int`
+            The number of expected lines/replies.
```
Comment (r-owen) [inline python/lsst/ts/ATHexapod/controller.py:125]: Consider combining has_response into num_line by using 0 for no responses.

This makes the API easier to understand in that the user doesn't have to wonder what happens if they specify has_response=False and num_line > 0.

File: python/lsst/ts/ATHexapod/controller.py [hunk 7]
```diff
@@ -117,28 +141,20 @@ def is_connected(self):
...
+                    replies = []
+                    for i in range(num_line):
+                        raw_line = await self.reader.readuntil(b'\n')
+                        line = raw_line.decode("ISO-8859-1").strip()
+                        self.log.debug(f"Read {i+1} of {num_line} lines: {line}")
+                        replies.append(line)
                 except asyncio.TimeoutError:
                     self.log.warning("Timed out waiting for response from controller. Result "
                                      "may be incomplete.")
+                    raise
```
Comment (r-owen) [inline python/lsst/ts/ATHexapod/controller.py:155]: Thank you for adding the raise!

Please consider changing the log message to a warning and removing "Result may be incomplete", because now no replies will be returned (there is no way to access the incomplete reply).

How did that work out in practice? Did the raise cause any problems for calling code?

**Review by r-owen — APPROVED** [review]
Looks very nice. I had a few minor suggestions which you can adopt or not as you see fit.

If you make changes, I'll leave it to you as to whether you want me to look at them. They're small enough that I doubt there's any need, but if you want a 2nd pair of eyes...

File: python/lsst/ts/ATHexapod/mock_server.py (hunk 4 above)
Comment (r-owen) [reply python/lsst/ts/ATHexapod/mock_server.py:304]: I don't understand your answer, and I wonder if you misunderstood my question. What I am asking is: does the real hexapod output data like the old string you had here (without newlines) or the new string (with newlines)?
